lib/mix/lib/mix/tasks/release.ex

```diff
@@ -0,0 +1,916 @@
+defmodule Mix.Tasks.Release do
+  @moduledoc """
+  Assembles a release for the current project:
+
+      MIX_ENV=prod mix release
+      MIX_ENV=prod mix release NAME
+
+  Once a release is assembled, it can be packaged and deployed to a
+  remote host, as long as the remote host runs on the same operating
+  system (OS) distribution and version as the machine running the
+  `mix release` command.
+
+  A release can be configured in your `mix.exs` file under the releases
+  key inside `def project`:
+
+      def project do
+        [
+          releases: [
+            demo: [
+              include_executables_for: [:unix],
+              applications: [runtime_tools: :permanent]
+            ],
+
+            ...
+          ]
+        ]
+      end
+
+  Each key represents the release name. Releasing a certain name is done
+  with:
+
+      MIX_ENV=prod mix release demo
+
+  If the given name does not exist, an error is raised.
+
+  If `mix release`, without a name, is invoked and there are multiple names,
+  the first name is used. If `mix release` is invoked and there are no names,
+  a release using the application name and default values is assembled.
+
+  ## Running the release
+
+  Once a release is assembled, you can start it by calling `bin/start`
+  inside the release. In production, you would do:
+
+      MIX_ENV=prod mix release
+      _build/prod/rel/my_app/bin/start
+
+  `bin/start` is a very short script that invokes the proper instruction
+  on `bin/RELEASE_NAME`. You can customize `bin/start` in any way you want,
+  but you are not supposed to change `bin/RELEASE_NAME`.
+
+  `bin/start` will start the system connected to the current standard
+  input/output, where logs are also written by default. That's the
+  preferred way to run the system. Many tools, such as `systemd`, platforms
+  as a service, such as Heroku, and many containers platforms, such as Docker,
+  are capable of processing the standard input/output and redirecting
+  the log contents elsewhere.
+
+  If you open up `bin/start`, you will also see there is an option to
+  run the release in daemon mode. In daemon mode, the system is started
+  on the background via [run_erl](http://erlang.org/doc/man/run_erl.html)
+  on Unix-like systems and [erlsrv](http://erlang.org/doc/man/erlsrv.html)
+  on Windows systems.
+
+  ## Deployments
+
+  It is very important that a release is assembled on a machine with the
+  same operating system (OS) version and tooling as the machine the system
```

Comment:
I think it is better to be more specific about the requirements here, specifically what requirements need to be met to build a release and deploy to a target. Something like the following:

A release is built on a **host**, a machine which contains Erlang, Elixir, and any other dependencies needed to compile your application. A release is then deployed to a **target**, potentially the same machine as the host, but usually separate, and often there are many targets (either multiple instances, or the release is deployed to heterogeneous environments).

To deploy straight from a host to a separate target without cross-compilation, the following must be the same between the host and the target:

- Target architecture (e.g. x86_64 vs ARM)
- Target Vendor+OS (e.g. Windows, Linux, Darwin/macOS)
- Target ABI (e.g. musl, gnu)

This is often represented in the form of target triples, e.g. `x86_64-unknown-linux-gnu`, `x86_64-unknown-linux-musl`, `x86_64-apple-darwin`. 

So to be more precise, to deploy straight from a host to a separate target, the Erlang Runtime System (ERTS), and any native dependencies (NIFs), must be compiled for the same target triple. If you are building on a MacBook (`x86_64-apple-darwin`) and trying to deploy to a typical Ubuntu machine (`x86_64-unknown-linux-gnu`), the release will not work. Instead you should build the release on a `x86_64-unknown-linux-gnu` host; trivially done by building on the target itself, but most commonly done by building the release in a virtual machine or container, which can be automated as part of your release process.

In addition to matching the target triple, it is also important that the target has all of the system packages that your application will need at runtime. A common one is the need for OpenSSL when building an application that uses `:crypto` or `:ssl`, which is dynamically linked to ERTS. The other common source for native dependencies like this comes from dependencies containing NIFs (natively-implemented functions) which may expect to dynamically link to libraries they use.

These system packages are typically managed using the system package manager, but if necessary, you can also bundle the compiled object files in the release, as long as they were compiled for the same target. If doing so, you need to update `LD_LIBRARY_PATH` with the paths containing the bundled objects.

Currently, there is no official way to cross-compile a release from one target triple to another, however it is possible to build ERTS on a given target once, copy it to your host, and then reference that "cross-compiled" ERTS rather than the host ERTS. However, this does not account for NIFs your release contains, which would need to be cross-compiled as well, and included in the release. Due to the complexity, it is generally recommended to use the approaches mentioned earlier if your target can be emulated in a virtual machine or container.

Reply:
This is a beautiful description, thank you. I have incorporated most of it into an explicit section called "Requirements".

Reply:
It could be possible to set this up similar to the rustup tool where you can download these target triples using something like `mix release.toolchain download x86_64-unknown-linux-gnu` and use something like `mix release.toolchain set x86_64-unknown-linux-gnu` to set prior to building your release, along with commands for updating, listing, deleting, etc. These target triples could be downloaded from hex similar to hex tooling itself. Just an idea.

Reply:
@beardedeagle the VM part is doable but the problem is also every NIF available which also needs to consider the Erlang runtime version if precompiled (so it becomes a quadruple and at this point probably too many possibilities). Others will be able explain it better than me. :)